README.md

# Standalone Cafe

This is a simple cafe software for Windows 10 PC allowing timing use of a Windows account using tickets with a range of durations.
User enters ticket number for a specific time of use and is locked out when ticket time expires.

## Setting Up
* Download the code to a preferred location and pip install modules in `requirements.txt`
* Run the `createTickets.py` file by doubleclicking or from cmd to create your tickets within the program directory. The tickets will be output to a file named `tickets.txt`
* To run the app, double click on the file `cafe.pyw`
* For better security, you can move the file `tickets.txt` to an inaccessible location so that others cant access the tickets. Also you can add `cafe.pyw` to startup programs so that the cafe starts when the user signs in to their account. For more security you can create a non-administrator account, disable task manager and control panel and auto start the cafe in that account.

## Contact
<a href="mailto:tsvenza@gmail.com">
Email Me</a> <br/>


changelog.txt
v1.6:
-TODO: Create gui for viewing and managing database efficiently
-improve dumpDB to dump with various modes
- adjust ticket to only add bonus for first one and add flat time for subsequent tickets
 eg first 10 min will add 13 mins but next will add 10 min to add up to 23 mins.
-create single click files for dumping database 
 and creating tickets

v1.5
-Add timestamps for datetime bought and datetime created for statistical purposes
-Removed string formatting to make database secure from injection attacks
-Edit utils.py to prevent closing of joystick configuration app x360ce if app is foreground.
 rather just minimize it.

v1.4
-Display hours, mins and seconds when time is greater than 60 mins. Previously only
 displayed mins and seconds
-Resize time left label font to to fit in window so text is not cut off

v1.3
-Keep centered on screen when not fullscreen and keep geometry
-Keep fullscreen geometry when showing as fullscreen. A bug sometimes happens that causes
 the window to change size when application closes another fullscreen application such
 as a game

v1.2
-close foreground fullscreen app when time is up so that keyboard is freed
-change setMaximized to setFullScreen so that app appears over taskbar when logged off
 or not time left
-some ui improvements including boldening font and increasing size in code editor

v1.1
-changed database from crude file based to sqlite3 based
-some ui improvements

v1.0
-initial version

utils.py


import subprocess, ctypes, os
from random import randint
import win32gui, win32con

import tawsql
        
bonus_time = tawsql.bonus_time #seconds added to first ticket to allow for loading of games and setting up

def get_running_apps():
    #Get the running apps on the system and return the list of names
    results = list()
    cmd = 'powershell "gps | where {$_.MainWindowTitle } | select Description, Id, Path'
    proc = subprocess.Popen(cmd, shell = True, stdout=subprocess.PIPE)
    for line in proc.stdout:
        if not line.decode()[0].isspace():
            print(line.decode().rstrip())
    
    
def minimise_running_app(handle_name):
    #mininise the app with the given names
    app_handle = ctypes.windll.user32.FindWindowW(None, handle_name)
    ctypes.windll.user32.ShowWindow(app_handle, 6)
    
def close_running_app(handle_name):
    #close the running app
    app_handle = ctypes.windll.user32.FindWindowW(None, handle_name)
    ctypes.windll.user32.CloseWindow(app_handle)
    
def minimise_foreground_app():
    app_name = win32gui.GetWindowText(win32gui.GetForegroundWindow())
    minimise_running_app(app_name)
    
def close_foreground_app():
    app_name = win32gui.GetWindowText(win32gui.GetForegroundWindow())
    #hack to prevent closing x360ce, just minimise_foreground_app
    if 'Jocys.com X360' in app_name:
        minimise_foreground_app()
        return
    app_handle = win32gui.FindWindow(None, app_name)
    win32gui.PostMessage(app_handle, win32con.WM_CLOSE, 0, 0)
    
def check_time_code(time_code):
    #checks if time code is valid.
    #returns 0 if code has been used already
    #returns -1 if code is invalid
    #returns time in seconds of the code
    try:
        time_code = int(time_code)
        return tawsql.checkCode(time_code)
    except Exception as e:
        print(e)
        return -1
        
def update_time_code(time_code, new_time, date_created=None, date_sold=None):
    #update the time code with new_time in seconds
    try:
        time_code = int(time_code)
        return tawsql.updateCode(time_code, new_time, date_created, date_sold)
    except Exception as e:
        print(e)
        return -1
        
def is_ticket_sold(time_code):
    return tawsql.isTicketSold(time_code)
    
def create_time_codes(seconds, quantity):
    #create time codes of the given quantity, each with the given seconds
    for x in range(quantity):
        code = randint(1000, 999999)
        tawsql.addCode(code, seconds)
        

debug = False
if debug == True:
    '''create_time_codes(600 + bonus_time, 30)
    create_time_codes(1200 + bonus_time, 30)
    create_time_codes(1800 + bonus_time, 30)
    create_time_codes(3600 + bonus_time, 30)'''
    #tawsql.deleteAllCodes()
    tawsql.dumpDB('dump.txt')
    pass
        
        

        

db_manager.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'db_manager.ui'
#
# Created by: PyQt5 UI code generator 5.15.1
#
# WARNING: Any manual changes made to this file will be lost when pyuic5 is
# run again.  Do not edit this file unless you know what you are doing.


from PyQt5 import QtCore, QtGui, QtWidgets


class Ui_DBManager(object):
    def setupUi(self, DBManager):
        DBManager.setObjectName("DBManager")
        DBManager.resize(800, 600)
        self.horizontalLayoutWidget = QtWidgets.QWidget(DBManager)
        self.horizontalLayoutWidget.setGeometry(QtCore.QRect(0, 0, 801, 41))
        self.horizontalLayoutWidget.setObjectName("horizontalLayoutWidget")
        self.horizontalLayout = QtWidgets.QHBoxLayout(self.horizontalLayoutWidget)
        self.horizontalLayout.setContentsMargins(0, 0, 0, 0)
        self.horizontalLayout.setObjectName("horizontalLayout")
        self.label = QtWidgets.QLabel(self.horizontalLayoutWidget)
        self.label.setObjectName("label")
        self.horizontalLayout.addWidget(self.label)
        self.tableView = QtWidgets.QTableView(DBManager)
        self.tableView.setGeometry(QtCore.QRect(0, 40, 800, 500))
        self.tableView.setObjectName("tableView")

        self.retranslateUi(DBManager)
        QtCore.QMetaObject.connectSlotsByName(DBManager)

    def retranslateUi(self, DBManager):
        _translate = QtCore.QCoreApplication.translate
        DBManager.setWindowTitle(_translate("DBManager", "Form"))
        self.label.setText(_translate("DBManager", "TextLabel"))


admin_db_gui.py

import sys, os
from PyQt5 import QtCore, QtGui, QtWidgets, QtSql
from PyQt5.QtCore import Qt

import tawsql
from db_manager import Ui_DBManager

script_loc = os.path.abspath(os.path.dirname(__file__))
print(script_loc)
os.chdir(script_loc)


class DBModel(QtSql.QSqlTableModel):
    
    def __init__(self):
        self.db = QtSql.QSqlDatabase.addDatabase('QSQLITE')
        self.db.setDatabaseName('data.db')
    
        super().__init__()
       
        self.setTable('codes_table')
        self.setEditStrategy(QtSql.QSqlTableModel.OnManualSubmit)
        self.select()
        
class MainWindow(QtWidgets.QMainWindow):
    def __init__(self):
        super().__init__()
        
        self.gui = Ui_DBManager()
        self.gui.setupUi(self)
        
        self.table = self.gui.tableView
        
        self.model = DBModel()
        self.table.setModel(self.model)
        
        
app = QtWidgets.QApplication(sys.argv)
window = MainWindow()
window.show()
app.exec_()

createTickets.py


import tawsql
import random


def _countUnusedTickets(cursor, countMap):
    #reset map
    for key in countMap:
        countMap[key] = 0
    #count tickets
    for ticket in cursor:
        key = ticket[tawsql.INITIAL_TIME_INDEX]
        if key in countMap: countMap[key] += 1
        
def _isMapFilled(countMap):
    for key in countMap:
        if countMap[key] < 30: return False
    return True
    
def _fillTicketsInDatabase(countMap):
    for time in countMap:
        count = 30 - countMap[time]
        for x in range(count):
            tawsql.addCode(random.randint(1000, 999999), time)
    
def createAndDumpTickets():
    #clear all used tickets first to clean db and keep it small
    print('Cleaning up database...')
    tawsql.deleteUsedCodes()
    #readd special code if it has been deleted
    tawsql.addCode(215454, 500000)
    print('Database cleaned')
    
    countMap = {tawsql.TEN: 0, 
                    tawsql.FOURTEEN: 0,
                    tawsql.TWENTY: 0,
                    tawsql.THIRTY: 0,
                    tawsql.HOUR: 0 }
    
    print('Filling database ...')
    s = '''SELECT * FROM CODES_TABLE WHERE (TIME!=0 AND TIME=INITIAL_TIME)'''
    cursor = tawsql.conn.execute(s)
    _countUnusedTickets(cursor, countMap)
    print(countMap)
    while not _isMapFilled(countMap):
        _fillTicketsInDatabase(countMap)
        cursor = tawsql.conn.execute(s)
        _countUnusedTickets(cursor, countMap)
    print('Filled database')
    print(countMap)
    print('Dumping tickets ...')
    tawsql.dumpDB('tickets.txt')
    print('Tickets dumped to Desktop folder')
    
    
if __name__ == '__main__':
    createAndDumpTickets()

cafe_time_main_ui.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'cafe_time_main_ui.ui'
#
# Created by: PyQt5 UI code generator 5.15.1
#
# WARNING: Any manual changes made to this file will be lost when pyuic5 is
# run again.  Do not edit this file unless you know what you are doing.


from PyQt5 import QtCore, QtGui, QtWidgets


class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        MainWindow.setObjectName("MainWindow")
        MainWindow.resize(816, 600)
        self.centralwidget = QtWidgets.QWidget(MainWindow)
        self.centralwidget.setObjectName("centralwidget")
        self.verticalLayoutWidget = QtWidgets.QWidget(self.centralwidget)
        self.verticalLayoutWidget.setGeometry(QtCore.QRect(81, 185, 471, 331))
        self.verticalLayoutWidget.setObjectName("verticalLayoutWidget")
        self.verticalLayout = QtWidgets.QVBoxLayout(self.verticalLayoutWidget)
        self.verticalLayout.setContentsMargins(0, 0, 0, 0)
        self.verticalLayout.setObjectName("verticalLayout")
        self.horizontalLayout = QtWidgets.QHBoxLayout()
        self.horizontalLayout.setSpacing(4)
        self.horizontalLayout.setObjectName("horizontalLayout")
        self.timeLeftLabel = QtWidgets.QLabel(self.verticalLayoutWidget)
        font = QtGui.QFont()
        font.setFamily("Segoe UI Black")
        font.setPointSize(16)
        font.setBold(True)
        font.setWeight(75)
        self.timeLeftLabel.setFont(font)
        self.timeLeftLabel.setText("")
        self.timeLeftLabel.setAlignment(QtCore.Qt.AlignCenter)
        self.timeLeftLabel.setWordWrap(False)
        self.timeLeftLabel.setObjectName("timeLeftLabel")
        self.horizontalLayout.addWidget(self.timeLeftLabel)
        self.verticalLayout.addLayout(self.horizontalLayout)
        spacerItem = QtWidgets.QSpacerItem(20, 40, QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Expanding)
        self.verticalLayout.addItem(spacerItem)
        self.label = QtWidgets.QLabel(self.verticalLayoutWidget)
        font = QtGui.QFont()
        font.setPointSize(12)
        self.label.setFont(font)
        self.label.setToolTip("")
        self.label.setTextFormat(QtCore.Qt.AutoText)
        self.label.setAlignment(QtCore.Qt.AlignBottom|QtCore.Qt.AlignHCenter)
        self.label.setObjectName("label")
        self.verticalLayout.addWidget(self.label)
        self.timeCodeEdit = QtWidgets.QLineEdit(self.verticalLayoutWidget)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Preferred)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(self.timeCodeEdit.sizePolicy().hasHeightForWidth())
        self.timeCodeEdit.setSizePolicy(sizePolicy)
        self.timeCodeEdit.setMinimumSize(QtCore.QSize(0, 0))
        font = QtGui.QFont()
        font.setPointSize(16)
        font.setBold(True)
        font.setWeight(75)
        self.timeCodeEdit.setFont(font)
        self.timeCodeEdit.setObjectName("timeCodeEdit")
        self.verticalLayout.addWidget(self.timeCodeEdit)
        self.horizontalLayout_2 = QtWidgets.QHBoxLayout()
        self.horizontalLayout_2.setObjectName("horizontalLayout_2")
        spacerItem1 = QtWidgets.QSpacerItem(40, 20, QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Minimum)
        self.horizontalLayout_2.addItem(spacerItem1)
        self.okButton = QtWidgets.QPushButton(self.verticalLayoutWidget)
        self.okButton.setObjectName("okButton")
        self.horizontalLayout_2.addWidget(self.okButton)
        spacerItem2 = QtWidgets.QSpacerItem(40, 20, QtWidgets.QSizePolicy.Expanding, QtWidgets.QSizePolicy.Minimum)
        self.horizontalLayout_2.addItem(spacerItem2)
        self.logoffButton = QtWidgets.QPushButton(self.verticalLayoutWidget)
        self.logoffButton.setObjectName("logoffButton")
        self.horizontalLayout_2.addWidget(self.logoffButton)
        self.verticalLayout.addLayout(self.horizontalLayout_2)
        self.horizontalLayout_3 = QtWidgets.QHBoxLayout()
        self.horizontalLayout_3.setObjectName("horizontalLayout_3")
        self.messagesLabel = QtWidgets.QLabel(self.verticalLayoutWidget)
        font = QtGui.QFont()
        font.setPointSize(10)
        font.setBold(True)
        font.setWeight(75)
        self.messagesLabel.setFont(font)
        self.messagesLabel.setText("")
        self.messagesLabel.setObjectName("messagesLabel")
        self.horizontalLayout_3.addWidget(self.messagesLabel, 0, QtCore.Qt.AlignHCenter)
        self.verticalLayout.addLayout(self.horizontalLayout_3)
        spacerItem3 = QtWidgets.QSpacerItem(20, 40, QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Expanding)
        self.verticalLayout.addItem(spacerItem3)
        MainWindow.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(MainWindow)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 816, 18))
        self.menubar.setObjectName("menubar")
        MainWindow.setMenuBar(self.menubar)
        self.statusbar = QtWidgets.QStatusBar(MainWindow)
        self.statusbar.setObjectName("statusbar")
        MainWindow.setStatusBar(self.statusbar)

        self.retranslateUi(MainWindow)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

    def retranslateUi(self, MainWindow):
        _translate = QtCore.QCoreApplication.translate
        MainWindow.setWindowTitle(_translate("MainWindow", "MainWindow"))
        self.label.setText(_translate("MainWindow", "<b>Enter Your Time Code</b>"))
        self.okButton.setText(_translate("MainWindow", "OK"))
        self.logoffButton.setText(_translate("MainWindow", "Log Off"))


tickets_db_dialog.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'tickets_db_dialog.ui'
#
# Created by: PyQt5 UI code generator 5.15.1
#
# WARNING: Any manual changes made to this file will be lost when pyuic5 is
# run again.  Do not edit this file unless you know what you are doing.


from PyQt5 import QtCore, QtGui, QtWidgets


class Ui_Dialog(object):
    def setupUi(self, Dialog):
        Dialog.setObjectName("Dialog")
        Dialog.resize(400, 300)
        self.buttonBox = QtWidgets.QDialogButtonBox(Dialog)
        self.buttonBox.setGeometry(QtCore.QRect(30, 240, 341, 32))
        self.buttonBox.setOrientation(QtCore.Qt.Horizontal)
        self.buttonBox.setStandardButtons(QtWidgets.QDialogButtonBox.Cancel|QtWidgets.QDialogButtonBox.Ok)
        self.buttonBox.setObjectName("buttonBox")
        self.horizontalLayoutWidget = QtWidgets.QWidget(Dialog)
        self.horizontalLayoutWidget.setGeometry(QtCore.QRect(0, 30, 401, 21))
        self.horizontalLayoutWidget.setObjectName("horizontalLayoutWidget")
        self.horizontalLayout = QtWidgets.QHBoxLayout(self.horizontalLayoutWidget)
        self.horizontalLayout.setContentsMargins(0, 0, 0, 0)
        self.horizontalLayout.setObjectName("horizontalLayout")
        self.ticketNumLabel = QtWidgets.QLabel(self.horizontalLayoutWidget)
        self.ticketNumLabel.setAlignment(QtCore.Qt.AlignCenter)
        self.ticketNumLabel.setObjectName("ticketNumLabel")
        self.horizontalLayout.addWidget(self.ticketNumLabel)
        self.timeLeftLabel = QtWidgets.QLabel(self.horizontalLayoutWidget)
        self.timeLeftLabel.setAlignment(QtCore.Qt.AlignCenter)
        self.timeLeftLabel.setObjectName("timeLeftLabel")
        self.horizontalLayout.addWidget(self.timeLeftLabel)
        self.initialTimeLabel = QtWidgets.QLabel(self.horizontalLayoutWidget)
        self.initialTimeLabel.setAlignment(QtCore.Qt.AlignCenter)
        self.initialTimeLabel.setObjectName("initialTimeLabel")
        self.horizontalLayout.addWidget(self.initialTimeLabel)
        self.dateCreatedLabel = QtWidgets.QLabel(self.horizontalLayoutWidget)
        self.dateCreatedLabel.setAlignment(QtCore.Qt.AlignCenter)
        self.dateCreatedLabel.setObjectName("dateCreatedLabel")
        self.horizontalLayout.addWidget(self.dateCreatedLabel)
        self.dateBoughtLabel = QtWidgets.QLabel(self.horizontalLayoutWidget)
        self.dateBoughtLabel.setAlignment(QtCore.Qt.AlignCenter)
        self.dateBoughtLabel.setObjectName("dateBoughtLabel")
        self.horizontalLayout.addWidget(self.dateBoughtLabel)
        self.scrollArea = QtWidgets.QScrollArea(Dialog)
        self.scrollArea.setGeometry(QtCore.QRect(0, 50, 401, 191))
        self.scrollArea.setWidgetResizable(True)
        self.scrollArea.setObjectName("scrollArea")
        self.scrollAreaWidgetContents = QtWidgets.QWidget()
        self.scrollAreaWidgetContents.setGeometry(QtCore.QRect(0, 0, 399, 189))
        self.scrollAreaWidgetContents.setObjectName("scrollAreaWidgetContents")
        self.scrollArea.setWidget(self.scrollAreaWidgetContents)

        self.retranslateUi(Dialog)
        self.buttonBox.accepted.connect(Dialog.accept)
        self.buttonBox.rejected.connect(Dialog.reject)
        QtCore.QMetaObject.connectSlotsByName(Dialog)

    def retranslateUi(self, Dialog):
        _translate = QtCore.QCoreApplication.translate
        Dialog.setWindowTitle(_translate("Dialog", "Tickets Database"))
        self.ticketNumLabel.setText(_translate("Dialog", "Ticket No."))
        self.timeLeftLabel.setText(_translate("Dialog", "Time Left"))
        self.initialTimeLabel.setText(_translate("Dialog", "Initial Time"))
        self.dateCreatedLabel.setText(_translate("Dialog", "Date Created"))
        self.dateBoughtLabel.setText(_translate("Dialog", "Date Bought"))


tawsql.py


import os, sqlite3, datetime

#seconds added to first ticket to allow for loading of games and setting up
bonus_time = 200 

#allowed initial times in seconds
TEN = 600 + bonus_time
FOURTEEN = 840 + bonus_time
TWENTY = 1200 + bonus_time
THIRTY = 1800 + bonus_time
HOUR = 3600 + bonus_time
fixed_times = [TEN, FOURTEEN, TWENTY, THIRTY, HOUR]

#indexes of data in database
TICKET_CODE_INDEX = 0
TIME_LEFT_INDEX = 1
INITIAL_TIME_INDEX = 2
DATE_CREATED_INDEX = 3
DATE_SOLD_INDEX = 4

db_name = 'data.db'
script_loc = os.path.abspath(os.path.dirname(__file__))
print(script_loc)
os.chdir(script_loc)
conn = sqlite3.connect(db_name, detect_types=sqlite3.PARSE_DECLTYPES |
                                             sqlite3.PARSE_COLNAMES)
print('database connected')

def createTimeCodesTable():
    conn.execute('''CREATE TABLE IF NOT EXISTS CODES_TABLE
        (TIME_CODE INT PRIMARY KEY NOT NULL,
        TIME INT NOT NULL,
        INITIAL_TIME INT NOT NULL,
        DATE_CREATED TIMESTAMP,
        DATE_SOLD TIMESTAMP);''')
    conn.commit()
        
def addColumnToCodesTable(columnName, columnType):
    s = 'ALTER TABLE CODES_TABLE ADD ? ?'
    data = (columnName,columnType)
    conn.execute(s, data)
    conn.commit()
    
def addCode(time_code, time, timestamp=None):
    '''add the code to database. timestamp will be set to current time if None'''
    #check if code exists
    s = 'SELECT TIME FROM CODES_TABLE WHERE TIME_CODE = ?'
    data_tuple = (time_code, )
    cursor = conn.execute(s, data_tuple)
    if cursor.fetchone() is None:
        #do if time_code doesnt already exist
        if timestamp == None:
            timestamp = datetime.datetime.now()
        s = '''INSERT INTO CODES_TABLE (TIME_CODE, TIME, INITIAL_TIME, DATE_CREATED) \
            VALUES (?, ?, ?, ?)'''
        data_tuple = (time_code, time, time, timestamp)
        conn.execute(s, data_tuple)
        conn.commit()
        return True
    else: return False #failed to add, code already exists
        
def getRow(time_code):
    '''returns None if row doesnt exist'''
    s = 'SELECT * FROM CODES_TABLE WHERE TIME_CODE = ?'
    data_tuple = (time_code,)
    cursor = conn.execute(s, data_tuple)
    return cursor.fetchone()
    
def getAllRows():
    '''returns cursor for all rows in database'''
    s = 'SELECT * FROM CODES_TABLE'
    cursor = conn.execute(s)
    return cursor

def isTicketSold(time_code):
    '''return true if date_sold is not None, false otherwise'''
    ticket = getRow(time_code)
    if ticket == None: return True
    elif ticket[DATE_SOLD_INDEX] == None: return False
    else: return True
    
def updateCode(time_code, new_time, date_created=None, date_sold=None):
    '''update time left on code'''
    s = 'UPDATE CODES_TABLE SET TIME = ? WHERE TIME_CODE=?'
    data_tuple = (new_time, time_code)
    conn.execute(s, data_tuple)
    if date_created != None:
        s = 'UPDATE CODES_TABLE SET DATE_CREATED = ? WHERE TIME_CODE=?'
        data_tuple = (date_created, time_code)
        conn.execute(s, data_tuple)
    if date_sold != None:
        s = 'UPDATE CODES_TABLE SET DATE_SOLD = ? WHERE TIME_CODE=?'
        data_tuple = (date_sold, time_code)
        conn.execute(s, data_tuple)
    conn.commit()
    
def deleteCode(time_code):
    s = 'DELETE FROM CODES_TABLE WHERE TIME_CODE = ?'
    data_tuple = (time_code,)
    conn.execute(s, data_tuple)
    conn.commit()
    
def deleteCodesWithTime(time):
    '''delete codes that have the given initial time'''
    s = 'DELETE FROM CODES_TABLE WHERE INITIAL_TIME = ?'
    data_tuple = (time,)
    conn.execute(s, data_tuple)
    conn.commit()
    
def deleteUsedCodes():
    '''Delete codes that have been used up.
    Delete codes that were bought but not used up in seven days since date of
        buying.
    '''
    #delete exhausted codes
    s = 'DELETE FROM CODES_TABLE WHERE TIME = 0'
    conn.execute(s)
    #delete old codes that were used but not exhausted
    s = "DELETE FROM CODES_TABLE WHERE TIME != INITIAL_TIME AND DATE_SOLD <= DATE('now', '-7 day')"
    conn.execute(s)
    conn.commit()
    
def exhaustCodesWithTime(time):
    '''use up codes that have the given initial time, ie set time left to 0'''
    s = 'UPDATE CODES_TABLE SET TIME = ? WHERE INITIAL_TIME=?'
    data_tuple = (0, time)
    conn.execute(s, data_tuple)
    conn.commit()
    
def checkCode(time_code):
    '''return -1 if code doesnt exist, return time left if code exists'''
    item = getRow(time_code)
    if item is None:
        print("No items")
        return -1
    else:
        time = item[TIME_LEFT_INDEX]
        return time
        
def dumpDB(file_name, mode=0):
    ''' mode=0 unused codes,
        mode=1 all codes,
        mode=2 unused codes,
    '''
    s = ''
    if mode == 0:
        s = '''SELECT * FROM CODES_TABLE WHERE (TIME!=0 AND TIME=INITIAL_TIME) 
            ORDER BY INITIAL_TIME ASC, TIME ASC'''
    elif mode == 1:
        s = 'SELECT * FROM CODES_TABLE ORDER BY INITIAL_TIME ASC, TIME ASC'
    elif mode == 2:
        s = '''SELECT * FROM CODES_TABLE WHERE (TIME!=0 AND TIME!=INITIAL_TIME) 
            ORDER BY INITIAL_TIME ASC, TIME ASC'''
    else: return
    cursor = conn.execute(s)
    with open(file_name, 'w') as f:
        f.write('\t Code : \t Time \t Time Left\n')
        prev_time = -1
        prices = {TEN: '15 bond\n',
                  FOURTEEN: '20 bond\n',
                  TWENTY: '30 bond\n',
                  THIRTY: '45 bond\n',
                  HOUR: '90 bond + bonus 10 minutes\n' }
        for item in cursor:
            time_left = (item[TIME_LEFT_INDEX]) // 60
            time = (item[INITIAL_TIME_INDEX] - 200) // 60
            if time != prev_time:
                #put newline separator when processing new time range
                f.write('\n')
                if item[INITIAL_TIME_INDEX] in prices:
                    f.write(prices[item[INITIAL_TIME_INDEX]])
                prev_time = time
            s = '\t {:>6d} \t {:} mins \t {:} mins\n'.format(item[0], time, time_left)
            f.write(s)
    

createTimeCodesTable()

